Support output descriptors in scantxoutset

sipa opened the pull request:
As promised, here is an implementation of my output descriptor concept (https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82) and integration within the `scantxoutset` RPC that was just added through #12196.

It changes the RPC to use descriptors for everything; I hope the interface is simple enough to encompass all use cases. It includes support for P2PK, P2PKH, P2WPKH, P2SH, P2WSH, multisig, xpubs, xprvs, and chains of keys - combined in every possible way.

sipa committed:
Add more methods to Span class

This introduces a rudimentary begin(), end(), operator[], and subspan to Span.

sipa committed:
Output descriptors module

sipa committed:
Add simple FlatSigningProvider

sipa committed:
Descriptor tests

sipa committed:
Swap in descriptors support into scantxoutset

sipa commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,615 @@
+#include <script/descriptor.h>
```
Fixed.

promag commented on src/script/sign.h:
```diff
@@ -28,6 +28,20 @@ class SigningProvider
+    bool GetCScript(const CScriptID& scriptid, CScript& script) const override;
+    bool GetPubKey(const CKeyID& keyid, CPubKey& pubkey) const override;
+    bool GetKey(const CKeyID& keyid, CKey& key) const override;
+
```
nit, remove empty line.

promag commented on src/script/sign.h:
```diff
@@ -28,6 +28,20 @@ class SigningProvider
+
+};
+
+FlatSigningProvider Merge(const FlatSigningProvider& a, const FlatSigningProvider& b);
```
At the moment this is only used in tests, move there?

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+        return true;
+    }
+};
+#endif
```
nit, add `// KEY_ORIGIN_SUPPORT`.

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+typedef std::vector<uint32_t> KeyPath;
+
+/** Interface for public key objects in descriptors. */
+struct PubkeyProvider {
```
nit, `{` in new line.

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+
+/** Interface for public key objects in descriptors. */
+struct PubkeyProvider {
+    virtual ~PubkeyProvider(){};
```
nit, `= default`?

promag reviewed (commented):
Few nits, it this for 0.17? Is `scantxoutset` considered API experimental?

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+    std::vector<Span<const char>> ret;
+    auto it = sp.begin();
+    auto start = it;
+    while (it != sp.end()) {
```
Add
```cpp
if (it == sp.end()) return ret;
```
for when string is empty.

sipa replied:
No, it should be `[""]`. Callers can rely on there always being at least one argument in the result.

promag replied:
In that case 👍

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+{
+    Span<const char> sp(descriptor.data(), descriptor.size());
+    auto ret = ParseScript(sp, ParseScriptContext::TOP, out);
+    if (sp.size() == 0 && ret) return std::move(ret);
```
Compiler warning:
```
script/descriptor.cpp:617:39: warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]
    if (sp.size() == 0 && ret) return std::move(ret);
                                      ^
script/descriptor.cpp:617:39: note: remove std::move call here
    if (sp.size() == 0 && ret) return std::move(ret);
                                      ^~~~~~~~~~  
```

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
-            "          }\n"
+            "2. \"scanobjects\"                  (array, required) Array of scan objects\n"
+            "    [                             Every scan object is either a string descriptor or an object:\n"
+            "        \"descriptor\",             (string, optional) An output descriptor\n"
```
Why give both options? `[{ "desc": "...", ...}, ...]` is enough?

promag replied:
> It's pretty verbose

On the client side I think it doesn't matter (unless it's a human making the call).

sipa replied:
Yes, the reason to support the short notation is to simplify things for humans.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+                desc_str = scanobject.get_str();
+            } else if (scanobject.isObject()) {
+                UniValue desc_uni = find_value(scanobject, "desc");
+                if (desc_uni.isNull()) throw JSONRPCError(RPC_INVALID_PARAMETER, "Descriptor needs to be provided in scan object");
```
Could have a test.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+                UniValue range_uni = find_value(scanobject, "range");
+                if (!range_uni.isNull()) {
+                    range = range_uni.get_int();
+                    if (range < 1 || range > 1000000) throw JSONRPCError(RPC_INVALID_PARAMETER, "range out of range");
```
Could have a test.

promag commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+    if (Func("pk", expr)) {
+        auto pubkey = ParsePubkey(expr, ctx != ParseScriptContext::P2WSH, out);
+        if (pubkey) return MakeUnique<SingleKeyDescriptor>(std::move(pubkey), P2PKGetScript, "pk");
+    }
```
Why not return `nullptr` in these cases?

promag commented on src/script/descriptor.h:
```diff
@@ -0,0 +1,100 @@
+/** Parse a descriptor string. Included private keys are put in out. Returns nullptr if parsing fails. */
+std::unique_ptr<Descriptor> Parse(const std::string& descriptor, FlatSigningProvider& out);
+
+#endif
```
```
#endif // BITCOIN_SCRIPT_DESCRIPTOR_H
```

sipa commented:
Rebased, removed future "key origin" support (not needed right now, ignore).

Also, I'd like people to first review the descriptor language itself (a very short summary is in the RPC help, more of a reference in script/descriptor.h, and a design document in https://gist.github.com/sipa/e3d23d498c430bb601c5bca83523fa82). In particular, I'm not convinced about the "old" function (which maps pubkeys to P2PK,P2PKH,P2WPKH, and P2SH-PWPKH) - it's very useful for dealing with existing Core wallets, but "old" doesn't really convey the right meaning. However, I don't see what else to use. "legacy" is confusing (as it would have a different meaning than legacy in -addresstype). "default" sounds like it could change over time (the intention is that it wouldn't). "sk" or "singlekey" sounds like it's encouraged and/or could change over time, ...

achow101 reviewed (commented):
Concept ACK

promag commented:
Regarding the "old" function, I think it could receive the version, for instance `default(args..., version)` — it should validate if args are compatible with the version specified (not useful at the moment).

sipa commented:
@promag There is no intention of ever adding another of those "defaults". Future versions should work with specific derivations (only P2WSH directly for example, rather than a collection).

sipa commented on src/script/descriptor.cpp:
```diff
@@ -0,0 +1,619 @@
+        CExtKey key;
+        if (!GetExtKey(arg, key)) return false;
+        out = EncodeExtKey(key);
+        for (auto entry : m_path) {
```
Fixed.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
             "        },\n"
-            "      ]\n"
+            "        ...\n"
+            "    ]\n"
```
Don't forget to add a command-line example here, it look me a while to figure out the syntax from reading the above.

I suggest an example with a typical BIP44 (legacy) xpub + BIP49 xpub would be helpful:
```
["pkh(legacy_xpub/0/*)", "pkh(legacy_xpub/1/*)", "sh(wpkh(segwit_xpub/0/*)", "sh(wpkh(segwit_xpub/1/*)"]
```

A second example could show the root master xpriv and then do hardened derivation of xpriv/44'/* or xpriv/49'/*

sipa replied:
Done.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -1984,67 +1985,36 @@ class CoinsViewScanReserver
+            "    addr(<address>)                      Outputs whose scriptPubKey corresponds to the specified address (does not include P2PK)\n"
+            "    raw(<hex script>)                    Outputs whose scriptPubKey equals the specified hex scripts\n"
+            "    old(<pubkey>)                        P2PK, P2PKH, P2WPKH, and P2SH-P2WPKH outputs for the given pubkey\n"
+            "    pkh(<pubkey>)                        P2PKH outputs for the given pubkey\n"
```
`sh(wpkh())` is probably quite common, so maybe add that here?

sipa replied:
Done.

Sjors reviewed (approved):
Concept ACK, and lightly tested on Ledger (P2SH-P2WPKH) and Blockchain (P2PKH) xpubs.

flack commented:
@sipa since you write that `old` is a "combination of P2PK, P2PKH, P2WPKH, and P2SH-P2WPKH for public key P.", you could call it `combo` instead. Or `ck` if you want it to look more like the other things that use `P`. And potentially prepend `bc` to indicate it's for backwards compatibility (but in that case someone should check urban dictionary to make sure `bcck` isn't some newfangled swear word :-))

sipa commented:
@flack Oh, I like that. Changed to "combo".

MarcoFalke commented:
Travis failure:

```
test_framework.authproxy.JSONRPCException: Invalid descriptor 'old(03abcaca414970a8d36624011628045855cb01cded780445a33688a0b905a47224)' (-5)

jonasschnelli commented:
First two findings:

* I think supporting `h` as equivalent to `'` in extended key derivation range would be great (JSON / shell / GUI). I suggest pulling in https://github.com/bitcoin/bitcoin/commit/203b4c94ce83981b0d76dd2adb778719ba868db1.

* I think the `IsHardened()` function is not correct. Using `0'/0'/*` will derive all hardened keys at the * level (same as `0'/0'/*'`). Seems to be currently impossible to derive unhardened key with the asterisk.

jonasschnelli committed:
[QA] Add xpub range tests in scantxoutset tests

jonasschnelli commented on src/rpc/blockchain.cpp:
```diff
@@ -2090,79 +2060,35 @@ UniValue scantxoutset(const JSONRPCRequest& request)
+                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Invalid descriptor '%s'", desc_str));
+            }
+            if (!desc->IsRange()) range = 1;
+            for (int i = 0; i < range; ++i) {
```
Seems like that results in an off-by-one issue (misses last key):
Suggest pulling fix: https://github.com/bitcoin/bitcoin/commit/875431149745cc517350c0620baa887c748d0b35

sipa replied:
BIP32 indexes are 0-based, so I think the current code is correct. If you expand 1000 keys, those will be indices 0 through 999.

jonasschnelli commented:
Additional xpub range tests can be pulled from here: https://github.com/bitcoin/bitcoin/commit/cc1ffb324b53855f7eae81a6b0df10bd677ca1a5

jonasschnelli commented:
@Sjors: A test for `'/*` would currently fail,... but it's simple to add (keys are already prepared in 
https://github.com/bitcoin/bitcoin/commit/cc1ffb324b53855f7eae81a6b0df10bd677ca1a5), just waiting for a fix for the hardened/non-hardened mix.

jonasschnelli committed:
Add experimental warning to scantxoutset

jonasschnelli commented:
Suggest pulling in the experimental warning for scantxoutset here: https://github.com/bitcoin/bitcoin/commit/3ec0958ea746d0db5edefff9f423cd2caab32857

sipa commented:
@jonasschnelli Good catch; the derivation for "/*" with a parent hardened step is incorrect. Fixing.

sipa committed:
[QA] Extend tests to more combinations

sipa commented:
@jonasschnelli Ok, fixed the derivation, changed the range to be max index rather than count, included your tests and experimental warning, and added a number of tests myself (generated with an independent BIP32 implementation).

sipa committed:
Support h instead of ' in hardened descriptor paths

sipa commented:
@jonasschnelli Added support for h instead of ' in hardened descriptor paths as well (plus some tests for it). Serializing a descriptor will always use '.

jonasschnelli reviewed (approved):
Tested ACK f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f

MarcoFalke commented:
Concept ACK f6b7fc3

laanwj commented:
utACK f6b7fc349ccf9cfbeb7e91e19c20e2a2fcc9026f

Sjors commented on test/functional/rpc_scantxoutset.py:
```diff
@@ -36,13 +52,39 @@ def run_test(self):
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/0)"])['total_amount'], Decimal("4.096"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1)"])['total_amount'], Decimal("8.192"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ "combo(tpubD6NzVbkrYhZ4WaWSyoBvQwbpLkojyoTZPRsgXELWz3Popb3qkjcJyJUGLnL4qHHoQvao8ESaAstxYSnhyswJ76uZPStJRJCTKvosUCJZL5B/1/1/1500)"])['total_amount'], Decimal("16.384"))
+        assert_equal(self.nodes[0].scantxoutset("start", [ {"desc": "combo(tprv8ZgxMBicQKsPd7Uf69XL1XwhmjHopUGep8GuEiJDZmbQz6o58LninorQAfcKZWARbtRtfnLcJ5MQ2AtHcQJCCRUcMRvmDUjyEmNUWwx8UbK/1/1/*')", "range": 1499}])['total_amount'], Decimal("1.536"))
```
Maybe use `range: 1` and `range: 2` everywhere to speed up these tests?
